### PR TITLE
feat: kms support rsa 4096 3072 2048

### PR DESCRIPTION
**Pull request by chengjoey**
#### What this PR does / why we need it:
kms support rsa 4096 3072 2048

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda-org.erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJpdGVyYXRpb24iOlsxMDkwXSwibWVtYmVyIjpbIjEwMDEyMDUiXX0%3D&id=282915&iterationID=1090&pId=0&type=TASK)

#### Specified Reviewers:

/assign @your-reviewer

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English | Support kms rsa 4096 3072 2048            |
| 🇨🇳 中文    |   kms的rsa           |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sfwn** [comment]
also support 2048/3072

**Comment by sfwn** [comment]
add your test images.

**Comment by chengjoey** [comment]
> add your test images.
1. create rsa key
image
 at the same time, will store the key info and version info
image

2. encrypt and decrypt:
image
image

3. rotate key version:
will add a new version in store and change key primary version
image

**Comment by chengjoey** [comment]
> also support 2048/3072

added

File: pkg/kms/plugins/dicekms/kms.go [hunk 1]
```diff
@@ -215,15 +246,36 @@ func (d *Dice) Decrypt(ctx context.Context, req *kmstypes.DecryptRequest) (resp
...
-		return nil, err
-	}
-	additionalDataJSON, err := json.Marshal(&additionalData{KeyID: keyInfo.GetKeyID()})
-	if err != nil {
-		return nil, err
+	var plaintextBytes []byte
+	switch keyInfo.GetKeySpec() {
+	case kmstypes.CustomerMasterKeySpec_ASYMMETRIC_RSA_4096, kmstypes.CustomerMasterKeySpec_ASYMMETRIC_RSA_2048,
+		kmstypes.CustomerMasterKeySpec_ASYMMETRIC_RSA_3072:
+		rsaCrypt := encryption.NewRSAScrypt(encryption.RSASecret{
```
Comment (sfwn) [inline pkg/kms/plugins/dicekms/kms.go:253]: NewRSA Scrypt is strange

**Review by sfwn — CHANGES_REQUESTED** [review]

**Comment by sfwn** [comment]
RSA has max data length limitation, so do you need check before encrypt? please test it.

**Comment by chengjoey** [comment]
> RSA has max data length limitation, so do you need check before encrypt? please test it.
The standard library `crypto` will check the length of src, and return error : `message too long`
image

**Comment by sfwn** [comment]
how about cipher type?

**Comment by chengjoey** [comment]
> how about cipher type?
CBC,CFB,OFB,ECB  ard different modes of aes encryption
image

File: pkg/kms/kmscrypto/rsa_2048_4096.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+	case Base64:
+		return base64.StdEncoding.EncodeToString(data), nil
+	case String:
+		return string(data), nil
+	default:
+		return "", fmt.Errorf("secretInfo OutputType unsupport")
+	}
+}
+
+type RsaCrypt struct {
```
Comment (sfwn) [inline pkg/kms/kmscrypto/rsa_2048_4096.go:75]: `RsaCrypt` `RSASceret`

**Review by sfwn — CHANGES_REQUESTED** [review]

File: pkg/kms/kmscrypto/rsa_2048_4096.go (hunk 2 above)
Comment (chengjoey) [reply pkg/kms/kmscrypto/rsa_2048_4096.go]: `RsaCrypt ` redefine as `RSACrypt`

File: pkg/kms/kmscrypto/rsa_2048_4096.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+//Secret defines the private key type
+type Secret uint
+
+const (
+	PKCS1 Secret = iota
+	PKCS8
+)
+
+//DecodeString decodes string data to bytes in designed encoded type
+func DecodeString(data string, encodedType Encode) ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (chengjoey) [reply pkg/kms/kmscrypto/rsa_2048_4096.go]: move these method `DecodeString` `EncodeToString` to util.go

**Review by sfwn — APPROVED** [review]

**Pull request merged**